pkg/controller/pvcontroller/sync.go

```diff
@@ -0,0 +1,195 @@
+				return errors.Wrapf(err, "failed to send delete event data of volume %s to server", pvObj.Name)
+			}
+
+			_, err = eventSender.AnnotateDeleteEvent(pvObj)
+			if err != nil {
+				return errors.Wrapf(err, "failed to annotate volume %s with delete event information", pvObj.Name)
+			}
+			klog.Infof("Successfully sent delete volume %s event to server", pvObj.Name)
+		}
+		err = eventSender.RemoveEventFinalizer()
+		if err != nil {
+			return errors.Wrapf(err, "failed to remove finalizers on volume %s", pvObj.Name)
+		}
+	}
+	return nil
+}
+
+func (pController *PVMetricsController) getEventSender(pvObj *corev1.PersistentVolume) (collectorinterface.EventsSender, error) {
+	var eventFinalizer string
+	for _, finalizer := range pvObj.Finalizers {
+		if strings.HasSuffix(finalizer, collectorinterface.OpenebsEventFinalizerSuffix) {
+			eventFinalizer = finalizer
+			break
+		}
+	}
+	if eventFinalizer == "" {
+		klog.Warningf("unable to find volume %s type from finalizers %v", pvObj.Name, pvObj.Finalizers)
+		return nil, nil
+	}
+
+	if value, ok := pvObj.Labels[nfspv.OpenEBSNFSLabelKey]; ok && value == "true" {
+		return tokenauth.NewTokenClient(
+			nfspv.NewNFSVolume(
+				pController.KubeClientset,
+				pController.PVCLister,
+				pController.PVLister,
+				pvObj,
+				pController.NFSServerNamespace)), nil
+	}
+	eventSenderType := eventFinalizer[:len(eventFinalizer)-len(collectorinterface.OpenebsEventFinalizerSuffix)]
+	return nil, errors.Errorf("event sender is not available for volume type %s", eventSenderType)
+}
+
+// canSkipEventCollection will return true based on following conditions:
+// 1. Retrun true if volume doesn't require event to be exported(based on annotation "events.openebs.io/required": "true")
+// 2. Retrun true if deletion timestamp is not set and create event already sent
+// 3. else return false
+func canSkipEventCollection(pvObj *corev1.PersistentVolume) bool {
+	if value, isExist := pvObj.Annotations[eventRequiredAnnotationKey]; !isExist || value != eventRequiredAnnotationValue {
+		return true
+	}
+	return pvObj.DeletionTimestamp == nil && isCreateVolumeEventSent(pvObj)
+}
+
+// isCreateVolumeEventSent will return true if volume has
+// suffix(event.openebs.io/volume-create) in annotations and value is sent
+func isCreateVolumeEventSent(pvObj *corev1.PersistentVolume) bool {
+	for key, value := range pvObj.Annotations {
+		if strings.HasSuffix(key, collectorinterface.OpenebsCreateAnnotationSuffix) {
+			return value == collectorinterface.OpenebsSentAnnotationValue
+		}
+	}
+	return false
+}
+
+// isDeleteVolumeEventSent will return true if volume has
+// suffix(event.openebs.io/volume-delete) in annotations and value is sent
+func isDeleteVolumeEventSent(pvObj *corev1.PersistentVolume) bool {
+	for key, value := range pvObj.Annotations {
+		if strings.HasSuffix(key, collectorinterface.OpenebsDeleteAnnotationSuffix) {
+			return value == collectorinterface.OpenebsSentAnnotationValue
+		}
+	}
+	return false
+}
```

Comment:
I think I should move these predicates to collector-interface(Which will make more sense).